File: bench/kernels/e3_index_arith/ref.c
/* e3-index-arith (family E), naive C. Scaled index arithmetic against an
 * opaque bound; `walk` is self-recursive so it does not inline in any of
 * the three languages and the bound stays a runtime value.
 * Protocol: argv[1]=ops; prints {"ns":..,"ops":..,"sink":..}. */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

static int64_t walk(int64_t n, int64_t depth) {
    int64_t acc = 0;
    for (int64_t i = 0; i < n; i++) acc = (acc + i * 8) & 1048575;
    if (depth > 0) acc = (acc + walk(n, depth - 1)) & 1048575;
    return acc;
}

int main(int argc, char **argv) {
    int64_t ops = argc > 1 ? (int64_t)strtoull(argv[1], 0, 10) : 2000;
    const int64_t inner = 100000;
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    int64_t sink = 0;
    for (int64_t k = 0; k < ops; k++) sink = (sink + walk(inner, 1)) & 1048575;
    clock_gettime(CLOCK_MONOTONIC, &t1);
    uint64_t ns = (uint64_t)(t1.tv_sec - t0.tv_sec) * 1000000000ull
                + (uint64_t)(t1.tv_nsec - t0.tv_nsec);
    printf("{\"ns\":%llu,\"ops\":%lld,\"sink\":%lld}\n",
           (unsigned long long)ns, (long long)(ops * inner * 2), (long long)sink);
    return 0;
}

File: bench/kernels/e3_index_arith/ref.rs
//! e3-index-arith (family E), Rust reference.
//! Protocol: argv[1]=ops; prints {"ns":..,"ops":..,"sink":..}.

use std::hint::black_box;

fn walk(n: i64, depth: i64) -> i64 {
    let mut acc: i64 = 0;
    let mut i: i64 = 0;
    while i < n {
        acc = (acc + i * 8) & 1048575;
        i += 1;
    }
    if depth > 0 {
        acc = (acc + walk(n, depth - 1)) & 1048575;
    }
    acc
}

fn main() {
    let ops: i64 = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(2000);
    let inner: i64 = 100_000;
    let t0 = std::time::Instant::now();
    let mut sink: i64 = 0;
    for _ in 0..ops {
        sink = (sink + walk(black_box(inner), 1)) & 1048575;
    }
    let ns = t0.elapsed().as_nanos();
    println!(
        "{{\"ns\":{ns},\"ops\":{},\"sink\":{sink}}}",
        ops.wrapping_mul(inner).wrapping_mul(2)
    );
}
